File: go/0001-two-sum.go
package main

// https://leetcode.com/problems/two-sum/description/

func twoSum(nums []int, target int) []int {

	indexMap := make(map[int]int, len(nums)-1)

	for i, el := range nums {
		indexMap[el] = i
	}

	for i, el := range nums {
		diff := target - el
		value, isExist := indexMap[diff]

		if isExist && value != i {
			return []int{i, value}
		}
	}

	return nil
}

// func main() {

// 	nums := []int{2, 7, 11, 15}

// 	fmt.Println(twoSum(nums, 9))
// }

File: rust/src/0001-two-sum.rs
use std::collections::HashMap;



pub fn two_sum(nums: Vec<i32>, target: i32) -> Vec<i32> {

    let mut index_map: HashMap<i32, usize> = HashMap::new();

    for (i, item) in nums.iter().enumerate(){

        let diff = target - item;

        if let Some(&j) = index_map.get(&diff) {
            return vec![i as i32, j as i32];
        }

        index_map.insert(*item, i);
    }

    vec![]
}

fn main() {
    let nums = vec![3,2,4];

    let target = 6;


    println!("{:?}", two_sum(nums, target));

}